### ISSUE TITLE
Lazy union (aka. no implicit union)

### ISSUE THREAD

**Issue by kintel** [openscad/openscad#350]
Don't perform a union in the following cases:
- Separate root objects
- `for` loops: A `for` loop will create a list of separate object passed on to the parent node (or the root)
- `children()`. A `children()` statement will pass on the list of children to the parent node

See also https://github.com/openscad/openscad/wiki/OEP2:-Implicit-Unions.

The code lives in https://github.com/openscad/openscad/tree/issue350

This feature is currently experimental. To enabled it, _lazy unions_ must be enabled in Preferences or the cmd-line (`--enable=lazy-union`).

Exporting top-level object to STL/OFF/AMF/DXF/SVG will attempt to keep the objects separate as far as the used file format supports it.

TODO:
- [x] Basic implementation
- [x] Tag as experimental feature (to be able to merge)
- [x] Test csg output (top level, `for()` and `children()`)
- [x] Test `intersection() for()`
- [x] Test `hull() for()`
- [x] Test nested for test
- [x] Test simple top-level objects
- [x] Test `minkowski() for()`
- [x] Test `difference() for()`
- [x] Test `children()`
- [x] STL export
- [x] OFF export
- [x] AMF export
- [x] DXF export
- [x] SVG export
- [x] Deprecate `intersection_for`: To combine this with the experimental feature, we need to defer builtin initialization until after preferences has been read => no need to combine it with experimental for now, deprecate `intersection_for` after we move out of the experimental stage
- [ ] Document the behavior of each file format for some slicer engines
- [ ] Consider the unwrap operator: https://github.com/openscad/openscad/wiki/OEP4%3A-Unwrap-Operator
- [ ] Evaluate compatibility with OpenSCAD2: https://github.com/doug-moen/openscad2/blob/master/rfc/Functions.md#fixing-the-module-composability-problem
- [ ] Test AMF import of separate objects.

This has been discussed a bit, and is interesting for a number of reasons:
- Speed: In many cases, most (half?) of the time and memory used for compiling an OpenSCAD design is consumed by the final union of the top-level objects (see e.g. #261 and #342)
- Necessity: Slicer software (Slic3r in particular, others unknown) supports slicing overlapping objects as it performs an implicit union of all submeshes. 
- Multi-material 3D printing: We could, at the top-level, specify different colors for different top-level objects. When exporting as separate meshes, these colors could be mapped to materials by the slicer software. Combined with amf or obj export, this could make multi-material prints a lot more convenient.
- Manifoldness: This will partially let users export more complex models where the final manifold test done today fails (see #337).

Further work past this issue:
- Don't need to perform _any_ unions which are not influenced by subsequent operations, independent on top-level or not. This will allow modules to create on non-unioned objects.
- Do lazy evaluation of CGAL Nef polyhedrons. This will make it possible to defer calculation of transformations until the last minute, which perhaps makes it possible to avoid caching more than one instance of geometries which only differ in transformation.
##

**Comment by bobc** [comment]
One could have the view that outputting overlapping STLs is a "hack", and AMF is the "proper" way to do multi-volumes.

In the case of AMF, the standard requires volumes to be non-overlapping.[1] So for AMF export, it would have to call back into the geometry functions before exporting to resolve the geometry. Exporting to AMF could then create an unexpected long delay, which currently is not cancellable. The AMF output might not match the output in other file formats.

That all seems rather messy. I think it would be much cleaner to have all the geometry resolved before export functions are called.

If we had a faster CSG, do we still need lazy unions?

[1] https://en.wikipedia.org/wiki/Additive_Manufacturing_File_Format#Structure

**Comment by kintel** [comment]
Lazy union is indeed a workaround for faster CSG. I foresee that CSG will still be a bottleneck, even if it's significantly faster, so we might eventually have to revisit this.

**Comment by francoisbecker** [comment]
Not sure it's the place or time for that, but what about a more simple "render by color()" and "export STL by color()", provided the scad file was using separate colors (e.g. color("red")…) for each part?

**Comment by GilesBathgate** [comment]
@francoisbecker I think the color() module only works for rendering using OpenCSG, and you cannot export a model to STL until it has been rendered in CGAL mode.

**Comment by francoisbecker** [comment]
@GilesBathgate Maybe I was not explicit enough: render all tree under a specific color using OpenCSG separately (and subsequent separate STL export with file naming according to the colour). Thanks for the link.

**Comment by GilesBathgate** [comment]
@francoisbecker What I am trying to get at is that the CGAL geometry doesn't contain color information. As I understand, (prior to this lazy union development) the entire geometry is compiled down into a single CGAL union node. There is simply not "separate things" to export as STL (yet). OpenCSG geometry is implicit and cannot be exported to STL.

**Comment by kintel** [comment]
@francoisbecker While exporting colors as separate files is interesting, you still need to guarantee that the exported STL files don't intersect, as AMF (and I assume slicers) require these volumes to be non-overlapping for 3D printing purposes. Having tools in OpenSCAD to help ensuring this would be cool.

Some notes about this here: https://github.com/openscad/openscad/wiki/Multi-material-support

I may have misunderstood your request :)

**Comment by GilesBathgate** [comment]
@kintel Checking whether the AABB overlaps is quite trivial, but for multi material probably not very useful (I assume people want to have two parts touching/interlocking) so I think you would have to use CGAL::Nef_polyhedron_3::intersect() operation and check to see if the result was the empty set via CGAL::Nef_polyhedron_3::is_empty()

**Comment by kintel** [comment]
@GilesBathgate exactly - checking might not be sufficient. We may want an operator which takes two volumes and returns two non-overlapping ones.

**Comment by GilesBathgate** [comment]
@kintel An operator that takes two overlapping volumes an returns to non-overlapping ones is just the difference operator, with the exception that you want to keep hold of the volume that was subtracted:

```
module volume1() {
difference() {
    cube(10);  
    cube(5);  //<-- keep a copy of this somewhere volume2
}
}

volume1();  //<--This is the other part.
```

**Comment by kintel** [comment]
@GilesBathgate yes, in essence. I guess it could be expanded into any number of volumes and done in user space.

**Comment by GilesBathgate** [comment]
@kintel Well yes you could do it in userspace, but I provided the openscad example to aid explanation (perhaps that was perhaps a bit misleading.)

I  meant that this new operation would be in C++:

```
GetOverlappingVolumes(
CGAL::Nef_polyhedron_3 inputVolume1,
CGAL::Nef_polyhedron_3 inputVolume2,
CGAL::Nef_polyhedron_3& outputVolume1,
CGAL::Nef_polyhedron_3& outputVolume2)
{
outputVolume1 = inputVolume1.difference(inputVolume2);
outputVolume2 = inputVolume2;
}
```

**Comment by thehans** [comment]
I was thinking about the backwards compatibility issue, and I was reminded of how javascript handles strict mode by placing a string on its own before any statements

    'use strict';
And how this can be used at the top of the file to cover the whole script, or at the top of individual functions for just the scope of that function.

A string by itself would not be not a valid statement in OpenSCAD, but what if it was implemented as a special $ variable that could be set at the top of any new script, or within specific modules, etc.  You could maybe also have OpenSCAD editor automatically put this line at the top of new blank scripts with a short comment.

    $implicit_union = true; // Recommeded for new scripts, see http://wiki/url for details

**Comment by kintel** [comment]
We've been debating such feature selection.  The challenge is that it gets messy really fast, when we keep adding new features.
The best two solutions so far, IMO, is either a version directive like '#openscad <version>" or implicitly determine version based on features used. Both are best suited for a major language upgrade though, not single features like this.

**Comment by oysteinkrog** [comment]
Any chance this could be merged?

**Comment by kintel** [comment]
See the TODO list for the missing pieces.
The main missing thing is evaluating forward compatibility with the current OpenSCAD2 proposal and either modifying this feature, modifying the OpenSCAD2 proposal, or document any incompatibility.

**Comment by oysteinkrog** [comment]
Do you think adding lazy unions for modules as well would be a lot of work?

**Comment by kintel** [comment]
I think it's just a matter of enabling support for modules, but it needs testing as well as evaluating potential backwards compatibility issues.

**Comment by oysteinkrog** [comment]
I wanted to work on this a bit but I had trouble getting it to work at all.
After finally getting a debug build up (so I could debug it properly) I've found that the Preferences::setupFeaturesPage() seems to override the -enable=lazy-union CLI parameter that I set...
I'm thinking that CLI parameters should probably override the serialized? settings.
I also don't see any features GUI at all, am I correct in assuming that's just not implemented?

**Comment by kintel** [comment]
Some notes:
* The Feature tab show up only if build using ```qmake CONFIG+=experimental```
* The cmd-line flags only affect cmd-line usage, the GUI settings only affect GUI usage

**Comment by patrickxia** [comment]
This doesn't seem to work in recursive modules. Here's an example where it doesn't seem to pass all the children to further invocations:

`
module arrange() {
  echo($children);
  if ($children <= 1) {
    children();
  } else {
    children(0);
    translate([10, 0, 0]) {
      arrange() children([1 : $children - 1]);
    }
  }
}

arrange() {
  cube(5);
  cube(5);
  cube(5);
}
`

Note the echos result in 

`
ECHO: 3
ECHO: 1
`

but we expect to see a call of the module with the size of children being two.

**Comment by patrickxia** [comment]
Ah, I see this is actually out of scope and the first "not in this issue" bullet point. Sorry about that. Consider this a vote for the module behavior being implemented as well.

**Comment by stuartpb** [comment]
> The best two solutions so far, IMO, is either a version directive like '#openscad "

I'm not sure if there's a better issue to discuss this idea in, but if you do end up going this route, I'd point out that OpenJSCAD is already using `//!OpenSCAD` as a directive at the top of a file to interpret code as being written for (the current version of) OpenSCAD.

**Comment by lostapathy** [comment]
I'd really like to experiment with this, but I'm having issues building it - I get issues very similar to https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=844881  Any ideas?

I hate to see us leave this much performance in a branch just because we're not sure how it mixes with the (probably far) future directions of OpenSCAD.  I feel like having this performance boost in play might change the whole narrative, for the better, about how to use OpenSCAD and what sort of applications it's capable of.

What if, instead of waiting until we can commit to a global change of behavior, we put this behind a system variable like $union.  When $union is true (the default) OpenSCAD would behave as it always has.  A user could then pass $union=false to modules where it's safe to do so and skip the implicit unions.

This would get us some real-world experience to better refine our vision for the OpenSCAD-2.  Older versions of OpenSCAD would just ignore $union, and only users who know to set $union=true would be affected by any potential problems.  If it all turns out to be mostly a non-issue (or we make it a non-issue by handling any unforeseen issues at export time), we can change the default to $union=false down the road so you get the performance by default.

**Comment by kintel** [comment]
@lostapathy I cherry-picked the fix for that issue from master and pushed it.

A good first step towards working on this would be to bring the branch in sync with master.
Using a `$union` variable and maintain backwards compatibility is a decent experiment. Getting it merged to master would be nice and allow us to more quickly iterate and seek external input.

**Comment by stuartpb** [comment]
How would `$union` interact with modules written to expect implicit unions, though?

**Comment by deckar01** [comment]
I built this branch and compared it to the `2017.08.03 (git cecf3b98)` build I use regularly. Rendering (F6) the chain ring example with this branch (feature on) ran in essentially 0s compared to ~30s on my "stable" build.

I ran [a brick benchmark](https://gist.github.com/deckar01/ef11def51de7e71d9f288c6e5819fdb7) that contains ~600 identical non-overlapping boxes. This branch (feature on) ran in ~1m20s compared to the ~60s on my "stable" build. Some of the additional time is likely attributable to the additional GCAL cache logging. Is this a valid test case for this feature branch?

**Comment by kintel** [comment]
@deckar01 This branch should only give significant speed-up if your object live on the top level, not if they're further down in the hierarchy. It should, however, not cause significant slowdown.

**Comment by kintel** [comment]
@deckar01 The primary challenge finalizing this is to figure out how to manage compatibility with existing designs. It shouldn't be hard to implement a non-union group. It might even be there as a switch already; I cannot remember the details.

**Comment by nrdvana** [comment]
What do you think about a flag on each op that previously used to imply a union, and then add a default for that flag in the parser?  This way anything created by the default parser will imply a union, and parsers that recognize they are reading new code can disable that flag?

```
for(x=[...]) {  // union = true
}
for(x=[...], union=false) {  // manually disabled
}
```
```
use 2.0  // borrow a notation from Perl.  Only affects this scope

use <lib/something.scad>   // will have its own scope, unaffected

for (x=[...]) { // union=false
}
```

**Comment by thehans** [comment]
Just updated this branch to be current with master.  It was only something like 2575 commits behind :stuck_out_tongue:

Seems to be working well with the current changes to just `children` and `for` loops.

I found that ListNode wasn't handling background/highlight modifiers and fixed that.

I suspect we want to eventually make all other potentially grouping nodes to use ListNode:
* if / else
* let (assign?)
* render
* echo / assert (related #3131)
* projection
* all transformations?
* linear_extrude, rotate_extrude?
* user defined modules
Not sure if I missed anything there.

Also still need to add 3mf support since that was added well after this was last worked on.

**Comment by MichaelAtOz** [comment]
> (assign?)

Perhaps it is time to drop deprecated things.

**Comment by thehans** [comment]
I agree Michael, not worth updating even if it would be simple to do so.  I would be fine with transitioning our currently deprecated modules and module parameters to be fully obsolete, not supported anymore.

Unrelated to that, there was some discussion on IRC of what difference it would make if we change change user-defined modules to act like ListNodes so I'm leaving my thoughts on that here.

ListNodes basically just append children to *their* parent node rather than grouping them in implicit union.  And in the case of user-defined modules I'm thinking about children from the *body of the module definition*, not children from module call/instantiation, which the builtin `children()` module would be used to access.

I have identified 3 cases where this change in behavior would affect the output:
1) If the module is called directly from the top level, or all its parent nodes up to root are also ListNodes, without any CSG operation in-between. 
2) If the module is called as the **first** child of a difference operation. 
`difference(){ union(){A(); B();} C(); } != difference() { A(); B(); C(); }`
3) if the module is **any** child of an intersection. 
`intersection() { A(); union(){B(); C();} } != intersection() { A(); B(); C(); }`

These could cause problems in terms of backwards compatibility, since I have a hunch that its fairly common for existing libraries or modules to rely on that implicit union to output a single object.  

Of course having the ability to define modules that for ex. output arrangements of disjoint geometries that don't require union, could also be very useful to speed up compile times.  

I'm picturing a case where a user wants to use the new features in their code, but also depends on an older library that wasn't written for it.  One solution could be to default to implicit union, but have some flag to enable/disable it?  Could be a special variable.  If this were done we'd also want to consider whether it should apply to just user-defined modules, or all the other potential list-able builtins too.

**Comment by thehans** [comment]
Branch updated again to add `echo` `assert` `let` and `if`.  
Also updated the highlight/background modifier handling again since I realized it was missing them in CSG dumps when applied to ListNodes, and fixed some other weird edge cases.

I enabled the lazy-union for all tests(change not committed) just to see what issues this might cause, and besides most dumptests failing (expected, because CSG dump has less `group` nodes now), there are just a handful of **Test failures**:
* `offcgalpngtest_issue1105b` and `offcgalpngtest_issue1215b` tests output blank images, so something wrong with .off export or import, I guess the format doesn't really support independent objects.  I wonder if meshes in OFF can overlap or maybe we need to force union for .off files.
* DXF export fails tests from overlapping or nested shapes.  I actually think allowing overlapping shapes could be really useful for laser engraving some kinds of patterns, as I've struggled with such issues in the past.  Maybe we should have an export option to union or not?
![dxf_overlap](https://user-images.githubusercontent.com/252233/69015384-e7443380-0958-11ea-8966-2fb1c426eeda.png)
![dxf_overlap2](https://user-images.githubusercontent.com/252233/69015385-ea3f2400-0958-11ea-9f0b-fcf0160bc81f.png)
* Failed `csgpngtest_2d-3d` and `cgalpngtest_2d-3d`.  F6 Render now shows 2D + 3D in the same view.  Need added check to fail in that case.  Results for both tests look same so only pasting one image:
![2Dplus3D](https://user-images.githubusercontent.com/252233/69015382-e3b0ac80-0958-11ea-9252-c323052192ef.png)

**Other bug noticed from manually testing**:
* Root modifier `!` on ListNodes needs fixing.  Testing on `for` loops gives blank Preview when there should be geometry, and crashes F6 Render.

**DIscussion of remaining modules to possibly convert to lazy union**:
* `render` - I think will be straightforward to implement, maybe I will try this next.
* **transformations** - a lot of different operations fall under this category, going by the cheatsheet's definition:
    * `translate`, `rotate`, `scale`, and `mirror` all reduce to `multmatrix` so just need to implement once for all those.  Not yet sure how much needs to change for this.  As @kintel says in the initial comment, this should be done lazily on CGAL Nef Polyhedrons, but will need to make it compatible with how we cache. We can compose nested transformations internally but I think we will want to keep the CSG dump output looking the same.
    * `color` - implementation will probably be similar to other matrix transformations since it can be store in State variable which is passed along through node visitor.
    * `resize` - should be do-able without union but will require some extra code to get individual bounding boxes, and extend those instead.
    * `offset` - does it matter if we lazy union this?
    * `hull` and `minkowski` - not applicable
* `projection` - Should we project 3D to 2D and keep all geometries separate?  Or project individually, then union the 2D objects?
* user-defined modules - any input on how we should handle these? (also see my prev comment)
* `linear_extrude` and `rotate_extrude` - these feel like lowest priority to me.  Do we even want them to lazy union at all?  2D union is fast, and I'm not sure there's much reason to allow for separate extrusion results.

**Comment by kintel** [comment]
Last I looked  at this, I ended up with a feature set that touched way too many parts, and still had the looming question of backwards compatibility. If you see any way of iterating our way to success, that would make this easier to merge.
Alternatively, if we end up having to (even temporarily) change behavior, tagging that as experimental would buy us time to figure out/refactor this in master rather than a long-living pull request.

**Comment by nrdvana** [comment]
Best way to solve back-compat is with a opt-in directive that affects *only* the current source file.  To borrow design from Perl,
```
use feature 'lazy_union';
```

or since 'feature' already means something in this codebase that can't be toggled on the fly (or can it?)
```
use options implicit_union=false;
```

Then if people decide it should be the default:
```
use 2020.xx;  // opt-in to default options as of version 2020.xx
```

**Comment by thehans** [comment]
Just realized a weird case that I thought would be a cool example, but doesn't work that way.
Giving a `for` loop as the child in a module call, results in `$children==1` rather than one for each iteration.

Not sure if this can be fixed without significant changes to how ModuleInstantiation and LocalScope work though.

```scad
// Note: Enable "View -> Animate"

// "particle" module with short name
module p() { square(); }

module explode() {
  echo($children);
  for(j=[0:1:$children-1]) {
    // angle is multiple of golden ratio to get "sunflower" pattern
    a = j*1.61803398875*360; 
    translate(j*$t*[sin(a),cos(a)]) children(j);
  }
}

// single child, does not move
explode() for(i=[0:1:200]) p();

// vs. specifying each child (works as intended, with or without lazy union)
/*
explode() {
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
  p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();p();
}
//*/
```

**Comment by nrdvana** [comment]
@thehans Here's the solution I came up with for that, though I don't know if it applies cleanly with this branch:

https://github.com/openscad/openscad/pull/2796/commits/f6cd34d8f172bcf663528261a63ff8c8157af64b

**Comment by thehans** [comment]
> Last I looked at this, I ended up with a feature set that touched way too many parts, and still had the looming question of backwards compatibility. If you see any way of iterating our way to success, that would make this easier to merge.

OK.  I fixed the 2d-3d mixed object issue, and the root modifier on ListNodes, and added support for multiple geometry export to 3MF.  I think the branch is actually in a reasonable state to merge as an experimental feature.  We can wait to implement lazy-union for other module types as separate PRs, if that's easier than waiting for one huge merge to be ready.  It would be nice to push this forward and start getting feedback from nightly users.

> Alternatively, if we end up having to (even temporarily) change behavior, tagging that as experimental would buy us time to figure out/refactor this in master rather than a long-living pull request.

Just to be clear, the current test setup on this branch passes 100%.  The other test failures I mentioned before were only when I temporarily applied the lazy-union feature to *every* regular test in the suite.  Additionally, these remaining failures are only when exporting non-manifold/intersecting geometry for DXF and OFF.  The OFF export succeeds in writing a file, and I can open them in meshlab, its just the import again into OpenSCAD which goes through CGAL functions and seems to silently fail for being non-manifold.

Some possible solutions:
1) Document in a release statement(eg. on the lists, and maybe in wiki somewhere) when we merge this, for users to be aware if they are exporting intersecting objects that they will need to call union themselves.  (Personally I think this option would be fine, since users who choose to enable an experimental feature should be expected to be aware of possible consequences)
2) Possibly add some check during export if any objects intersect and print a warning?  Not sure if we can check this more efficiently than just performing a union.
3) Force union on exports even when lazy-union feature is enabled - this is the least desirable option in my opinion, as it defeats the purpose of these changes.  And would make for less flexible usage, such as the example of purposely exporting DXF with intersecting objects for laser engraving, where they represent tool paths rather than 'solids".

**Comment by kintel** [comment]
@thehans Sounds good. I'll review this and push towards merging. Thanks for taking this on and reviving this ancient branch! Would you mind opening a PR, so we can discuss further details there?

**Comment by nophead** [comment]
Shouldn't export to STL always do an implicit union if objects overlap?

**Comment by deckar01** [comment]
One use case for preserving overlapping solids in a single STL is multi-material slicers. The layout of the parts is preserved during a single import, but the components can be configured independently with different colors/materials.

**Comment by nophead** [comment]
If they share vertices the result will not be a valid STL and probably won't import properly. Use another format if you want to do that.

Otherwise this will create a whole new wave of broken STLs on Thingiverse where people forgot to union the top level.

**Comment by deckar01** [comment]
This is already being done by other applications. The majority of complex models I find on thingiverse contain overlapping solids, slicers handle them fine, but OpenSCAD does not. I need to check the file contents, but I assumed they were declaring multiple solids in the STL file.

**Comment by nophead** [comment]
Not all slicers handle them fine. Skeinforge definitely doesn't.

With multiple materials they shouldn't overlap but if they touch that is enough to break STL. If they are separated by a micron they would not need to be unioned.

**Comment by kintel** [comment]
@nophead Yeah, slicers do vary. My checkbox above "Document the behavior of each file format for some slicer engines" was meant as an activity to build some knowledge around this.
We should probably break such discussion up into separate tickets once we re-establish a reasonable list of missing pieces related to this.

**Comment by thehans** [comment]
I do like the idea from [nrdvana's suggestion](https://github.com/openscad/openscad/issues/350#issuecomment-554807479) re: backwards compatibility.

Something that requires a scad file to declare when it relies on a new, experimental, or optional feature, rather than(or in addition to) GUI based preferences which will differ across users.  This makes it clear what features need to be supported, and the examples given would at least "fail fast" as a parser error on older releases, rather than output some possibly incorrect geometry while acting as if everything is fine.

I'm a bit hesitant to merge after thinking about this.  Should we implement such a check before we end up with ever more mix of scripts in the wild which aren't clear about which features or OpenSCAD version they were written for?

**Comment by kintel** [comment]
I think merging is fine. We already have the experimental system where all features are turned off by default, and that has served us well so far.
..but starting on a more well-defined way of managing this would definitely be welcome

**Comment by t-paul** [comment]
We may need that kind of definition for actual release features that might a compatibility mode. I think for experimental stuff it's the wrong solution. This is intended to change, potentially in incompatible ways while it's experimental, so putting it into designs will get messy.

**Comment by thehans** [comment]
OK, I went ahead and merged.   I will open a separate issue for discussing solutions to compatibility issues going forward.

**Comment by thehans** [comment]
Found another highlight & background modifier issue from testing this complex assembly: https://github.com/CarlosGS/Cyclone-PCB-Factory 

I ended up simplifying the issue down to this `previewChildren` example which is supposed to make it so children only show during preview, not render:
```
module previewChildren() { %children(); }
$fn=20;
previewChildren() 
  minkowski() { 
    translate([-5,-5]) cube(); 
    translate([5,5]) cylinder();
  }
translate([-5,5]) sphere(2);
```

I thought I had solved the issue of modifiers on ListNodes by passing these flags via the `State &state` parameter of `NodeVisitor::visit(...)`.  But this ends up applying to all children recursively, when we actually need to apply it to just the *immediate* children of a ListNode.

So during preview, the above example ends up showing the minkowski result as a background obejct, but also the individual minkowski components backgrounded by themselves.  F6 Render output is fine though.

`State` already had these highlight and background flags defined but I noticed they weren't being used in code anywhere.  Maybe because it doesn't work!  

I'm not sure a good way around this because child nodes are basically stored as a list of pointers to `const AbstractNode`.  
`AbstractNode` class doesn't store modifier flags and even if we added that, its still const.
Maybe we can apply the modifiers directly to the `ModuleInstantiation` objects of the ListNode children instead?

CSG dumps of above example:
(without lazy union)
```
group() {
%  group() {
    minkowski(convexity = 0) {
      multmatrix([[1, 0, 0, -5], [0, 1, 0, -5], [0, 0, 1, 0], [0, 0, 0, 1]]) {
        cube(size = [1, 1, 1], center = false);
      }
      multmatrix([[1, 0, 0, 5], [0, 1, 0, 5], [0, 0, 1, 0], [0, 0, 0, 1]]) {
        cylinder($fn = 20, $fa = 12, $fs = 2, h = 1, r1 = 1, r2 = 1, center = false);
      }
    }
  }
}
multmatrix([[1, 0, 0, -5], [0, 1, 0, 5], [0, 0, 1, 0], [0, 0, 0, 1]]) {
  sphere($fn = 20, $fa = 12, $fs = 2, r = 2);
}
```
(with lazy union)
```
group() {
%  minkowski(convexity = 0) {
%    multmatrix([[1, 0, 0, -5], [0, 1, 0, -5], [0, 0, 1, 0], [0, 0, 0, 1]]) {
%      cube(size = [1, 1, 1], center = false);
    }
%    multmatrix([[1, 0, 0, 5], [0, 1, 0, 5], [0, 0, 1, 0], [0, 0, 0, 1]]) {
%      cylinder($fn = 20, $fa = 12, $fs = 2, h = 1, r1 = 1, r2 = 1, center = false);
    }
  }
}
multmatrix([[1, 0, 0, -5], [0, 1, 0, 5], [0, 0, 1, 0], [0, 0, 0, 1]]) {
  sphere($fn = 20, $fa = 12, $fs = 2, r = 2);
}
```

**Comment by MichaelAtOz** [comment]
There is also a problem with modifiers and render.
It is also strangely only with some color-schemes.
Change the scheme to Tomorrow Night. 
Disable Lazy Union.
```
translate([-5,0]) cube(3);
difference() {
    translate([5,0]) cube(3);
    %cube(2);
}
```
Remove the `%`.
F6;

![Capture lazy colour](https://user-images.githubusercontent.com/1972961/70688286-81ec2400-1d05-11ea-85fd-3da5f98a06da.JPG)

Enable Lazy Union, Flush Cache, F6;

![Capture lazy colour b](https://user-images.githubusercontent.com/1972961/70688388-c4156580-1d05-11ea-9a8f-4f7f49d49a9f.JPG)

Reinsert the `%`, F6;

![Capture lazy colour c](https://user-images.githubusercontent.com/1972961/70688452-f030e680-1d05-11ea-9a13-7e494e1a45b0.JPG)

I've got no idea what's going on.

**Comment by hansonry** [comment]
dd

**Comment by hansonry** [comment]
Hey people!

If it helps motivate anyone, I could use this feature right about now. I would like to import my mesh into an application that only takes convex hulls. I can make that happen with a bunch of cubes, but the auto-union keeps the mesh concave.

Just so you know. I am rooting for you to succeed!

**Comment by thehans** [comment]
@hansonry The feature is already available as experimental (must be enabled in preferences) if you use a "nightly" build or basically any dev snapshot after Nov 18, 2019 when the code was merged.
Give it a try and see if it works for your use-case.

**Comment by s-ol** [comment]
I have been playing with it a bit for the purpose of laser cutting and engraving, and I agree that it would be nice if `translate()` etc., custom `module()`s' top-level scope and possibly also `group()` would be lazy as well. For example it would be very useful to define a bunch of overlapping 2D shapes for engraving within a module, then replicate that module in a for loop and be able to export an SVG or DXF with those outlines.

**Comment by fancsali** [comment]
I just started to use OpenSCAD a few days ago, and realised, I'd need this feature badly. I like the composability of modules and such a lot.

So, just wondering, is there a roadmap to promote this from and experimental to a core feature?

**Comment by wbu** [comment]
Since this issue is still open, I'll comment here instead of opening a new one.
What is the rationale behind the decision to make this a configuration option? If I want to distribute a .scad file to other people, I now have to put a union around every for loop, to make sure, that my scad file works for everybody, independent of their settings.
Is there any way to configure this behavior inside a scad file? What about doing this in a non-breaking way, for example by adding a new keyword like `ifor` for "inline for" or something similar?

I am not just nitpicking, it's a real issue for me, see https://github.com/wbu/drawer_organizer/issues/4

**Comment by t-paul** [comment]
It's not an option, it's flagged as experimental feature, meaning it will automatically be disabled for release builds. As "experimental feature" hopefully implies, it's not ready to be relied upon.

**Comment by wbu** [comment]
According to https://stackoverflow.com/questions/52219941/openscad-drill-into-group-created-by-for/53777838#comment114637501_55287051 at least Ubuntu seems to be shipping a non release build. Also in my openscad-nightly on debian this "feature option" is available.
If you did not built it yourself, the option looks not very much experimental :)

![image](https://user-images.githubusercontent.com/1161016/113929406-54779380-97f0-11eb-9b1c-50d2b964626d.png)

**Comment by t-paul** [comment]
Yes, nightly build = development snapshot = experimental features enabled to allow gathering feedback on those features. I don't know which Ubuntu version you refer to, but we have no control over what they ship. I would assume they ship the Debian build and that is a release build.

**Comment by wbu** [comment]
I see. Thank you. In the mean time, I could narrow my issue down. It does not seem to be related to lazy-union at all. Don't get me wrong: I really like the lazy-union functionality, just not the current implementation as a flag.

I am sorry for the noise.

**Comment by t-paul** [comment]
So we can't remove this experimental feature logic, it's the only way to get feedback from people who can't or just don't want to build OpenSCAD themselves. But I do see how it can create issues when running OpenSCAD from form sort of automation which may not have control over the exact version. Do you control the application invocation? In that case it might be a possibility to add something like `--disable-experimental` which would override any setting for experimental features in GUI and command line (and would be ignored in release versions).

**Comment by nrdvana** [comment]
I'll re-suggest that there should be a flag inside of the openscad file with file-scope.  So, with the feature available in the application, then a .scad file can request to enable that feature within the scope of one file.  That way there are no surprises and trying to use a scad file in an application version without that feature will fail with an easy-to-diagnose error message.

**Comment by Lenbok** [comment]
I'm confused by this feature. The tests have a case like this:

```
points=600;
random_angles=rands(0,360,points,42);

hull()
  for(i=[0:3:points-1])
    rotate([random_angles[i],random_angles[i+1],random_angles[i+2]]) translate([10,0,0]) cube();
```

It outputs a solid ball type thing. I would have thought with this feature, I could replace the call to `hull` with the commonly employed `hullchain`, i.e.:
```
points=600;
random_angles=rands(0,360,points,42);

module hullchain() {
  assert($children > 1, "hullchain requires multiple children");
  for (i = [0:$children - 2]) {
    hull() {
      children(i);
      children(i + 1);
    }
  }
}

hullchain()
  for(i=[0:3:points-1])
    rotate([random_angles[i],random_angles[i+1],random_angles[i+2]]) translate([10,0,0]) cube();
```
And the result should be a hulling of consecutive points, ending up with more of a string ball effect. But that is not what happens - the assertion fires instead, warning that there is only one child. Why does `intersection() for()` apparently work when lazy unions are enabled, but `hullchain() for()` doesn't? Am I doing something wrong?

**Comment by nophead** [comment]
I think `for` or another module can still only return one child regardless of lazy union.

**Comment by Lenbok** [comment]
Why should `intersection` any different to `hullchain` in this example? It certainly seems inconsistent! I would suggest that the currently implemented version of lazy union is incorrect if these two are treated differently.

**Comment by nophead** [comment]
I didn't realise `intersection() for()` had changed. It seems it now does the same as `intersection_for()` when lazy union is enabled but modules other than `for()` still seem to do an implicit union. For example translate([0,0]). So it seems `for` is now a special case.

I agree it is inconsistent but I don't think it is possible to count the number of children early enough when they are generated by `for()`, so I can't see how it will ever be fixed.

**Comment by baikal** [comment]
Late to the party - but the 'hullchain' use case is exactly what I was hoping for a long time with that initial announcement of lazy union.

So how can `for()` pass the number and 'list' of children up to intersection/difference/minkowski, but not to another (user) module? So `for()` "knows" by which module it is called and does something differently?

**Comment by sdasda7777** [comment]
Hi, I can't get this feature to work in the snapshot 2022.06.26 for Windows. But it might be me misunderstanding how to use it (I have checked the checkbox in the settings and have three separate top-level objects), or me exporting to bad formats (tried .obj and .stl). Any ideas?

**Comment by MichaelAtOz** [comment]
Lazy union does not include and mechanism for material definition. It is a prerequisite to be able to so so.

**Comment by sdasda7777** [comment]
> Lazy union does not include and mechanism for material definition. It is a prerequisite to be able to so so.

Of course, I am aware those are distinct features. However as the lazy union is currently not observable for me, I was hoping someone could with more knowledge could guess where the issue lies.

**Comment by MichaelAtOz** [comment]
(caveat - I don't know what it does when two objects overlap, for STL I assume it will be bad)

STL wont be different it is just a triangle soup. it just does not spend time doing a union.
3MF with lazy, you get (unzip it, see `3dmodel.model`)
```
...
	<resources>
		<object id="1" name="OpenSCAD Model" type="model" p:UUID="8470f547-3012-4d40-91c5-b22f4f397709">
			<mesh>
				<vertices>
					<vertex x="0" y="0" z="0" />
```
and `object id="2"` etc.

You should also check Help/Library-info, you should see the enabled feature flagged`*` e.g.
`Features: ... , input-driver-dbus, lazy-union*, vertex-object-renderers, ...`

**Comment by jordanbrown0** [comment]
[ In the comments below, multi-material includes multi-color. ]

> Lazy union does not include and mechanism for material definition. It is a prerequisite to be able to so so.

I'm dubious that lazy union will play a significant part in multi-material handling.  I think it'll go much deeper than that, and will not require lazy union per se.  (Not, I perhaps should note, that I fully understand lazy union.  But it's been discussed as avoiding the top-level union, and I think multi-material support will need to go deeper than that, and doesn't actually require lazy union for any one material.)

In particular, I think you should be able to do this:
```
    difference() {
        union() {
            color("red") cube(10);
            translate([0,0,10]) color("blue") cube(10);
        }
        translate([5,5,0]) cylinder(h=60, d=5, center=true);
    }
```
and get sensible results.  The color has to persist through both the inner union and the difference.

Lazy union, as I understand it, might help if the model separates its top-level objects by material, but if you're going to make the model do that then it's already 95% of the way to working today - just need to iterate over the materials producing individual STLs.  That seems like a very weak definition of multi-material support.

**Comment by sdasda7777** [comment]
@MichaelAtOz It looks exactly like you describe it. That must mean the issue is with the export to .obj. Thank you so much.

**Comment by nophead** [comment]
Can I ask what exactly the problem is. I design machines with hundreds of parts that are separate objects but the implicit union never gets in the way because I never hit F6 on an assembly of separate objects. I only use F6 to make STL files of one object at a time. With the F5 preview union is a NOP. It just draws each object in the specified colour.

**Comment by phryk** [comment]
Well, I think there's several things playing into this for me:

One is my pretty stubborn refusal to make cutouts any bigger than
they're actually supposed to be, which always leads to z-fighting in
spades – unless I actually render the assembly. I recently added
a `cutout_overcommit` variable to my desk project to avoid this,
but am only doing so under duress. :P

Another is having this more as a dependency for a feature I want:
Automated part extraction and generation of sawing/drilling/milling
patterns. This I think is most of the technical meat of the issue for
me.

A third one is just plain confusion, since there doesn't really seem to
be much in terms of established patterns concering devops/software
engineering for OpenSCAD projects. If you got any hints here, I'd be
delighted to hear them – especially if you managed to suss out how to
get FEM analysis for projects designed with OpenSCAD going. :3

**Comment by t-paul** [comment]
I'm not sure if that's what you mean, but for OpenSCAD CI, maybe have a look at https://openflexure.org/ .

Everything else does not seem to fit here as lazy union is not really intended to have much user visibility except for maybe faster processing and/or fixing a couple of awkward workarounds (like `intersection_for()`).

**Comment by nophead** [comment]
I don't understand your issue with cutouts. why does it matter if the length of a hole is slightly more than the material it goes through? Do you output the dimensions of the cylinder that drills a hole? Engineering drawings will detail the hole, not the length of the drill that makes it.

To avoid z-fighting when parts come near to each other in an assembly, I render any 3D differences in parts that are not STLs but not necessarily the whole object. But I never render objects that are separate, so I never get an issue with implicit union in preview. When making STLs I need any parts that overlap unioning into a solid and disjoint parts are not affected by union, other than GGAL is slow doing nothing. So I generally render an STLs that have differences and the union is no problem.

I make most of my parts in my library from unions of extruded 2D cross sections and avoid as much 3D differencing as possible.

To check if a screw fits an hole I tend to use the clip() module from my library to cross section the object, or just part of it.

![image](https://user-images.githubusercontent.com/566149/196971528-b7d0dc1e-fb83-4b8c-a915-c9a9895b9da2.png)

My only use of transparency is for transparent objects like this.

![main_assembled](https://user-images.githubusercontent.com/566149/196973081-fb9876c7-26e1-40ed-8117-fb9fef3f909c.png)

And then I need to arrange the order in which they are added to the assembly correctly for the viewing direction. It would be nice if OpenSCAD sorted them by Z when it drew them.

I agree OpenSCAD isn't great if you want drawings. I produce assembly diagrams with exploded views but all my parts are made by CNC of one sort or another rather than manual tools. 

![x_end_left_assembly](https://user-images.githubusercontent.com/566149/196975740-db44299d-8696-4891-b349-c28eb59bfc50.png)

For the original MDF version of Mendel90 I did output PDF drill templates that I spray mounted on the MDF that had cross hairs on the holes and their sizes.

![image](https://user-images.githubusercontent.com/566149/196975356-44a7b12f-8b1c-42f5-b396-9b3857b5cf74.png)

**Comment by jordanbrown0** [comment]
Getting rid of implicit union doesn't really mean anything to the user if the export mechanisms still export single-object formats.  This effort is primarily an optimization effort, not a functionality effort.

What you really want is a mechanism to define multiple parts that can be exported into multiple files, or into a single file that supports multiple distinct objects.

You want a mechanism that lets
```
color("red") cube(10,center=true);
color("blue") sphere(d=13);
```
yield something that can be separated into its red and blue components.  (Or, mostly equivalently, into its PLA and PVA components, or its steel and wood components.)

For those topics, see #1608, #1041, and [Multi material support](https://github.com/openscad/openscad/wiki/Multi-material-support).

**Comment by coryrc** [comment]
For anyone else not following the mailing list:

> Why would lazy union will never be a part of a release?

[It's broken and there is no fix. Some parts are obviously useful, so
other and more direct ways need to be found getting those features
without breaking things.](https://lists.openscad.org/empathy/thread/OJU22SNGU7EGLJICS5GCNFWM3DB6D4Y6?hash=QQUJZUTNG4JODSACOTSULWZWA6I7OGJX#QQUJZUTNG4JODSACOTSULWZWA6I7OGJX)

**Comment by coryrc** [comment]
Is anybody working on just preserving removal of the root union so [colorscad](https://github.com/jschobben/colorscad) can work?

**Comment by coryrc** [comment]
> Is anybody working on just preserving removal of the root union so [colorscad](https://github.com/jschobben/colorscad) can work?

I just checked and the root union is gone when exporting to 3mf.

So, given it's said lazy union is not going to be kept, can this Issue be closed? Or repurposed as "remove lazy union experiment and code"?

**Comment by jordanbrown0** [comment]
Top-level union sure seems to be done when I export to 3MF.   `cube(10); cube(10, center=true);` yields a shape with no [0,0,0] vertexes.

**Comment by t-paul** [comment]
It does for me (OpenSCAD Version: 2025.10.13.nightly (git c167e59) on Debian/Testing):
```
unzip -p /tmp/c.3mf 3D/3dmodel.model
<?xml version="1.0" encoding="utf-8"?>
<model xmlns="http://schemas.microsoft.com/3dmanufacturing/core/2015/02" unit="millimeter" xml:lang="en-US" xmlns:m="http://schemas.microsoft.com/3dmanufacturing/material/2015/02" xmlns:p="http://schemas.microsoft.com/3dmanufacturing/production/2015/06" xmlns:b="http://schemas.microsoft.com/3dmanufacturing/beamlattice/2017/02" xmlns:s="http://schemas.microsoft.com/3dmanufacturing/slice/2015/07">
	<metadata name="Application">OpenSCAD (https://www.openscad.org/)</metadata>
	<metadata name="CreationDate">2025-10-15T15:53:58Z</metadata>
	<resources>
		<m:colorgroup id="3">
			<m:color color="#3E999FFF" />
		</m:colorgroup>
		<object id="1" name="OpenSCAD Model 1" type="model" p:UUID="28d1d1b6-c401-458e-a26c-9719e078ef9d" pid="3" pindex="0">
			<mesh>
				<vertices>
					<vertex x="0" y="0" z="0" />
					<vertex x="0" y="0" z="10.000000" />
					<vertex x="0" y="10.000000" z="0" />
					<vertex x="0" y="10.000000" z="10.000000" />
					<vertex x="10.000000" y="0" z="0" />
					<vertex x="10.000000" y="0" z="10.000000" />
					<vertex x="10.000000" y="10.000000" z="0" />
					<vertex x="10.000000" y="10.000000" z="10.000000" />
				</vertices>
				<triangles>
					<triangle v1="0" v2="1" v3="3" />
					<triangle v1="0" v2="2" v3="6" />
					<triangle v1="0" v2="3" v3="2" />
					<triangle v1="0" v2="4" v3="5" />
					<triangle v1="0" v2="5" v3="1" />
					<triangle v1="0" v2="6" v3="4" />
					<triangle v1="1" v2="5" v3="3" />
					<triangle v1="2" v2="3" v3="6" />
					<triangle v1="3" v2="5" v3="7" />
					<triangle v1="3" v2="7" v3="6" />
					<triangle v1="4" v2="6" v3="5" />
					<triangle v1="5" v2="6" v3="7" />
				</triangles>
			</mesh>
		</object>
		<object id="2" name="OpenSCAD Model 2" type="model" p:UUID="01e01857-6dc7-40fa-8e96-dfd45902eb30" pid="3" pindex="0">
			<mesh>
				<vertices>
					<vertex x="-5.000000" y="-5.000000" z="-5.000000" />
					<vertex x="-5.000000" y="-5.000000" z="5.000000" />
					<vertex x="-5.000000" y="5.000000" z="-5.000000" />
					<vertex x="-5.000000" y="5.000000" z="5.000000" />
					<vertex x="5.000000" y="-5.000000" z="-5.000000" />
					<vertex x="5.000000" y="-5.000000" z="5.000000" />
					<vertex x="5.000000" y="5.000000" z="-5.000000" />
					<vertex x="5.000000" y="5.000000" z="5.000000" />
				</vertices>
				<triangles>
					<triangle v1="0" v2="1" v3="3" />
					<triangle v1="0" v2="2" v3="6" />
					<triangle v1="0" v2="3" v3="2" />
					<triangle v1="0" v2="4" v3="5" />
					<triangle v1="0" v2="5" v3="1" />
					<triangle v1="0" v2="6" v3="4" />
					<triangle v1="1" v2="5" v3="3" />
					<triangle v1="2" v2="3" v3="6" />
					<triangle v1="3" v2="5" v3="7" />
					<triangle v1="3" v2="7" v3="6" />
					<triangle v1="4" v2="6" v3="5" />
					<triangle v1="5" v2="6" v3="7" />
				</triangles>
			</mesh>
		</object>
	</resources>
	<build p:UUID="c4788fb0-7e15-457a-8fee-ea91992cbb24">
		<item objectid="1" partnumber="Part 1" p:UUID="0ac9b0df-6ef8-4e0c-b21f-14c981278b33" />
		<item objectid="2" partnumber="Part 2" p:UUID="c7376ead-dfc9-4154-84bd-f7a852b35fd1" />
	</build>
</model>
```

**Comment by jordanbrown0** [comment]
Is that with lazy union on, or off?  I interpreted Cory's comment as saying that the top-level union doesn't happen when lazy-union is **off**.  What you're getting looks like lazy-union **on**.

**Comment by coryrc** [comment]
> Is that with lazy union on, or off? I interpreted Cory's comment as saying that the top-level union doesn't happen when lazy-union is **off**. What you're getting looks like lazy-union **on**.

Well, I'm confused now, because I can export a 3mf and go into Orca and choose "break to parts" and it worked. I thought that was blocked by top-level union? (also [someone on reddit thought it was working](https://www.reddit.com/r/openscad/comments/1o4okmn/mask_curve_in_openscad/nj3ys5q/))

Hmm, maybe it worked because the objects weren't touching. Trying the following with CGAL and Manifold:

```
cube(10);
cube(10, center=true);
```

```
cube(10);
color("green") cube(10, center=true);
```

Nope, none can be separated. Sorry, tricked by Orca figuring out things aren't touching and letting them split. But thank you for showing me how to check if it did work!

I did look at the 3mf output and it did appear as one object, so I was confused how it knew. But now I know.

**Comment by coryrc** [comment]
So, are the steps to close this issue:

1. Remove implicit top-level union (from at least Manifold?)
2. Remove lazy union code

?

**Comment by jordanbrown0** [comment]
Removing the implicit top-level union is kind of the opposite of removing lazy union.  (Or did you mean those as alternatives?)

The language is defined to do a union of the top level.  I would consider it to be a backward compatibility error if the two overlapping cubes that I showed were **not** melded into one obvious polyhedron.

I think it's important to consider the intent.  I believe the intent of lazy union was a performance improvement, rather than being intended to directly support multi-color/multi-material models.  I think a multi-material solution needs to do something sensible with unions, so that `color("blue") cube(10); color("green") cube(10, center=true);" yields a blue volume and a green volume that are disjoint.  I think it's too much of an imposition to require that the model avoid overlapping colors; there are too many cases where the intent is clear but it's awkward to ensure that the two don't touch.  And that multi-material behavior should apply top-to-bottom; it should not matter how deeply buried the two-color subassembly is.

```
difference() {
    union() {
        color("blue") cube(10);
        color("green") cube(10, center=true);
    }
    sphere(6);
}
```
should yield a model with a green part and a blue part, disjoint.  (I have no strong opinion on what color the overlapping cube(5) is.  It needs to be well-defined, but could be either blue or green.  I have a mild preference for blue, for "first wins", because that's the way some of the other operations seem like they should work.)

**Comment by coryrc** [comment]
> Removing the implicit top-level union is kind of the opposite of removing lazy union. (Or did you mean those as alternatives?)

I did not. The only feature of lazy-union I want is to make https://github.com/jschobben/colorscad work, which seems to require the top-level not be unioned.

>  I think a multi-material solution needs to do something sensible with unions, so that `color("blue") cube(10); color("green") cube(10, center=true);" yields a blue volume and a green volume that are disjoint.

There's discussion on this and I think it's too complicated to be resolved anytime soon. But removing the implicit top-level union is fairly simple and really simple for people to preserve the old behaviour if it's needed for something (`union() {` and `}`). It won't affect STL output because it must be unioned to output, right? So it can only affect the newer export formats and I'm not sure why it would be worse to do.

**Comment by jordanbrown0** [comment]
I don't think the STL exporter knows how to do unions, nor should it.  Current STL export of a lazy-unioned pair of overlapping cubes does not union them; it yields an invalid STL with internal faces.

I have no opinion on whether exporters that support multiple parts should divide disjoint pieces of the model into multiple parts.

**Comment by coryrc** [comment]
> I did not. The only feature of lazy-union I want is to make https://github.com/jschobben/colorscad work, which seems to require the top-level not be unioned.

Err, I thought that, but I don't see it on the page. I always forget what thing I'm doing needs lazy union, until I load it :(.
